File: components-registry-service-light-client/src/main/java/org/octopusden/octopus/components/registry/light/client/dto/ErrorResponse.java
package org.octopusden.octopus.components.registry.light.client.dto;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * {@code ignoreUnknown} is mandatory here: the decoder is constructed with a
 * caller-supplied {@link com.fasterxml.jackson.databind.ObjectMapper}, and a
 * strict mapper would otherwise throw on any field added to the server's error
 * body (e.g. {@code errorCode}) — silently degrading typed 404/425 exceptions
 * to generic Feign errors (the decoder swallows the parse failure).
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public class ErrorResponse {
    private final String errorMessage;
    private final String errorCode;

    @JsonCreator
    public ErrorResponse(@JsonProperty("errorMessage") String errorMessage,
                         @JsonProperty("errorCode") String errorCode) {
        this.errorMessage = errorMessage;
        this.errorCode = errorCode;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    /**
     * Machine-readable error class (nullable; absent on older servers).
     * Known values: OPTIMISTIC_LOCK, UNIQUENESS_VIOLATION, DATA_INTEGRITY.
     */
    public String getErrorCode() {
        return errorCode;
    }
}

File: components-registry-service-core/src/main/kotlin/org/octopusden/octopus/components/registry/core/dto/ErrorResponse.kt
package org.octopusden.octopus.components.registry.core.dto

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty

/**
 * `errorCode` is a machine-readable discriminator for clients that must react
 * differently to error classes sharing one HTTP status — e.g. the Portal needs
 * to distinguish an optimistic-lock 409 ("reload and re-apply") from a
 * uniqueness-violation 409 ("fix the conflicting value"). Nullable: handlers
 * that have no meaningful class emit plain messages, and older servers omit
 * the field entirely. Known values: `OPTIMISTIC_LOCK`, `UNIQUENESS_VIOLATION`,
 * `DATA_INTEGRITY` — clients must tolerate unknown values.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
data class ErrorResponse(
    @JsonProperty("errorMessage") val errorMessage: String,
    // NON_NULL is wire-compat-bearing: v1-v3 ERROR bodies are byte-diffed against the baseline, and
    // an emitted `"errorCode": null` diverges on every error case. Null = omit the field.
    @JsonInclude(JsonInclude.Include.NON_NULL)
    @JsonProperty("errorCode") val errorCode: String? = null,
)
